File: mbeps1/ReadFiles/Python2/pflreadf1.f90
!-----------------------------------------------------------------------
! This program reads real periodic 1d fluid data
! written for 1D OpenMP PIC codes
! written by Viktor K. Decyk, UCLA
      program pflreadf1
      use in1, only: idrun, indx, ndim, dt
      use cmfield1
      use graf1
!
      implicit none
      integer, parameter :: ns = 2
      integer :: iudm = 19, iuv = 12
      integer :: i, n, m, ii, nx, npro, nprd, nrec, ierr
      integer :: nplot = 1, nts = 0
      real :: time
! nscalars = table of available diagnostics
      integer, dimension(ns) :: nscalars = 0
      real, dimension(:,:), allocatable :: fms, vfield, ufield
      real, dimension(:), allocatable :: sfield
      character(len=20), dimension(ns) :: dname = (/                    &
     &'elect fluid moments ','ion fluid moments   '/)
      character(len=8), dimension(ns) :: sname = (/'ELECTRON','ION     '&
     &/)
      character(len=16), dimension(5) :: ename = (/' DENSITY        ',  &
     &' VELOCITY FIELD ',' PRESSURE TENSOR',' ENERGY         ',         &
     &' HEAT FLUX      '/)
      character(len=10) :: cdrun
      character(len=32) :: fname
!
      write (*,*) 'enter idrun:'
      read (5,*) idrun
      write (cdrun,'(i10)') idrun
      cdrun = adjustl(cdrun)
      fname = 'diag1.'//cdrun
      call ffopen1(iudm,fname)
!
! determine which fluid diagnostics are available
      call readfldiags1(iudm,nscalars)
!
! select diagnostic
      m = sum(nscalars)
      if (m > 1) then
         n = -1
         do while (n < 0)
            do i = 1, ns
               if (nscalars(i)==1) then
                  write (*,*) 'enter ', i, 'for ', trim(dname(i))
               endif
            enddo
            read (5,*) n
            if (n==0) stop
            if ((n >= 1).and.(n <= ns)) then
               if (nscalars(n)==0) n = -1
            else
               n = -1
            endif
            if (n < 0) then
               write (*,*) 'invalid entry, try again or enter 0 to quit'
            endif
         enddo
      else if (m==1) then
         do i = 1, ns
            if (nscalars(i)==1) then
               n = i
               exit
            endif
         enddo
      else
         write (*,*) 'no scalar diagnostic files found'
         stop
      endif
!
      write (*,*) trim(dname(n)), ' diagnostic selected'
!
! return parameters for selected fluid diagnostic
      call fldiagparams1(iudm,n,nts,npro,nprd,nrec,fname)
      nplot = npro
!
! nx = number of global grid points in x direction
      nx = 2**indx
!
! allocate vector arrays
      allocate(fms(nprd,nx),sfield(nx))
      if (ndim==3) then
         allocate(vfield(ndim,nx))
         if (npro > 2) allocate(ufield(2*ndim,nx))
      endif
      dt = dt*real(nts)
!
! open stream file for vector field
      call fsopen1(iuv,fname)
!
! nrec = number of complete records
      write (*,*) 'records found: nrec = ', nrec
!
! open graphics device
      ierr = open_graphs(nplot)
!
! read and display data
      do ii = 1, nrec
! read real vector field
         call freadv1(iuv,fms,nprd,nx)
         time = dt*real(ii - 1)
! electrostatic case
         if (ndim==1) then
! display density in real space
            if (npro > 0) then
               sfield = fms(1,:)
               call dscaler1(sfield,trim(sname(n))//trim(ename(1)),ii,  &
     &999,1,nx,ierr)
               if (ierr==1) exit
            endif
! display velocity field in real space
            if (npro > 1) then
               sfield = fms(2,:)
               call dscaler1(sfield,trim(sname(n))//trim(ename(2)),ii,  &
     &999,0,nx,ierr)
               if (ierr==1) exit
            endif
! display pressure tensor in real space
            if (npro > 2) then
               sfield= fms(3,:)
               call dscaler1(sfield,trim(sname(n))//trim(ename(3)),ii,  &
     &999,0,nx,ierr)
               if (ierr==1) exit
            endif
! display electron heat flux in real space
            if (npro==4) then
               sfield = fms(5,:)
               call dscaler1(sfield,trim(sname(n))//trim(ename(5)),ii,  &
     &999,0,nx,ierr)
               if (ierr==1) exit
            endif
! electromagnetic case
         else if (ndim==3) then
! display density in real space
            if (npro > 0) then
               sfield = fms(1,:)
               call dscaler1(sfield,trim(sname(n))//trim(ename(1)),ii,  &
     &999,1,nx,ierr)
               if (ierr==1) exit
            endif
! display velocity field in real space
            if (npro > 1) then
               vfield  = fms(2:4,:)
               call dvector1(vfield,trim(sname(n))//trim(ename(2)),ii,  &
     &999,0,2,nx,ierr)
               if (ierr==1) exit
            endif
! display pressure tensor in real space
            if (npro > 2) then
               ufield= fms(5:10,:)
               call dvector1(ufield,trim(sname(n))//trim(ename(3)),ii,  &
     &999,0,2,nx,ierr)
               if (ierr==1) exit
            endif
! display electron heat flux in real space
            if (npro==4) then
               vfield = fms(12:14,:)
               call dvector1(vfield,trim(sname(n))//trim(ename(5)),ii,  &
     &999,0,2,nx,ierr)
               if (ierr==1) exit
            endif
         endif
      enddo
!
      call closeff(iudm)
      call closeff(iuv)
      call close_graphs()
!
      end program

File: mbeps1/ReadFiles/Python2/pflreadf1.py
#-----------------------------------------------------------------------
# This program reads real periodic 1d vector data
# written for 1D OpenMP PIC codes
# written by Viktor K. Decyk, UCLA
import sys
import math
import numpy

sys.path.append('./ReadFiles')
from libcmfield1 import *
from fgraf import *

int_type = numpy.int32
double_type = numpy.float64
float_type = numpy.float32
complex_type = numpy.complex64

ns = 2
iudm = 19; iuv = 12
nplot = 1
dname = numpy.array(["elect fluid moments ","ion fluid moments   "],
                     'S20')
sname = numpy.array(["ELECTRON","ION     "],'S8')
ename = numpy.array([" DENSITY        "," VELOCITY FIELD ",
                     " PRESSURE TENSOR"," ENERGY         ",
                     " HEAT FLUX      "],'S16')

# create string from idrun
idrun = int(input("enter idrun: "))
cdrun = str(idrun)
fname = "diag1." + cdrun
cmfield1.ffopen1(iudm,fname)

# nscalars = table of available diagnostics
nscalars = numpy.zeros((ns),int_type,'F')

# determine which fluid diagnostics are available
cmfield1.readfldiags1(iudm,nscalars)

# select diagnostic
m = numpy.sum(nscalars)
if (m > 1):
   n = -1
   while True:
      if (n < 0):
         for i in xrange(0,ns):
            if (nscalars[i]==1):
               print "enter ", i+1," for ", numpy.str.rstrip(dname[i])
         n = int(input(""))
         if (n==0):
            exit(1)
         if ((n >= 1) and (n <= ns)):
            if (nscalars[n-1]==0):
               n = -1
         else:
            n = -1
         if (n > 0):
            break
         print "invalid entry, try again or enter 0 to quit"
elif (m==1):
   for i in xrange(0,ns):
      if (nscalars[i]==1):
         n = i + 1
         break
else:
   print "no scalar diagnostic files found"
   exit(1)

print numpy.str.rstrip(dname[n-1]), " diagnostic selected"

nts = numpy.zeros((1),int_type,'F')
npro = numpy.zeros((1),int_type,'F')
nprd = numpy.zeros((1),int_type,'F')
mrec = numpy.zeros((1),int_type,'F')
fname = numpy.array([""],'S32')

# return parameters for selected fluid diagnostic
cmfield1.fldiagparams1(iudm,n,nts,npro,nprd,mrec,fname)
nrec = mrec[0]
nplot = npro[0]

# nx = number of global grid points in x direction
nx = int(math.pow(2,in1.indx))

# allocate vector arrays
fms = numpy.empty((nprd[0],nx),float_type,'F')
sfield = numpy.empty((nx),float_type,'F')
if (in1.ndim==3):
   vfield = numpy.empty((in1.ndim,nx),float_type,'F')
   if (npro[0] > 2):
      ufield = numpy.empty((2*in1.ndim,nx),float_type,'F')
dt = in1.dt*float(nts[0])

# open stream file for vector field
cmfield1.fsopen1(iuv,fname)

# nrec = number of complete records
print "records found: nrec = ", nrec

ierr = numpy.zeros((1),int_type,'F')
# open graphics device
ierr[0] = graf1.open_graphs(nplot)

# read and display data
for ii in xrange(0,nrec):
# read real vector field
   cmfield1.freadv1(iuv,fms,nprd[0],nx)
   time = dt*float(ii)
# electrostatic case
   if (in1.ndim==1):
# display density in real space
      if (npro[0] > 0):
         sfield[:] = fms[0,:]
         fname = numpy.str.rstrip(sname[n-1])+numpy.str.rstrip(ename[0])
         graf1.dscaler1(sfield,fname,ii,999,1,nx,ierr)
         if (ierr[0]==1):
            exit(1)
# display velocity field in real space
      if (npro[0] > 1):
         sfield[:] = fms[1,:]
         fname = numpy.str.rstrip(sname[n-1])+numpy.str.rstrip(ename[1])
         graf1.dscaler1(sfield,fname,ii, 999,0,nx,ierr)
         if (ierr[0]==1):
            exit(1)
# display pressure tensor in real space
      if (npro[0] > 2):
         sfield[:] = fms[2,:]
         fname = numpy.str.rstrip(sname[n-1])+numpy.str.rstrip(ename[2])
         graf1.dscaler1(sfield,fname,ii,999,0,nx,ierr)
         if (ierr[0]==1):
            exit(1)
# display electron heat flux in real space
      if (npro[0]==4):
         sfield[:] = fms[4,:]
         fname = numpy.str.rstrip(sname[n-1])+numpy.str.rstrip(ename[4])
         graf1.dscaler1(sfield,fname,ii,999,0,nx,ierr)
         if (ierr[0]==1):
            exit(1)
# electromagnetic case
   elif (in1.ndim==3):
# display density in real space
      if (npro[0] > 0):
         sfield[:] = fms[0,:]
         fname = numpy.str.rstrip(sname[n-1])+numpy.str.rstrip(ename[0])
         graf1.dscaler1(sfield,fname,ii,999,1,nx,ierr)
         if (ierr[0]==1):
            exit(1)
# display velocity field in real space
      if (npro[0] > 1):
         vfield[:,:] = fms[1:4,:]
         fname = numpy.str.rstrip(sname[n-1])+numpy.str.rstrip(ename[1])
         graf1.dvector1(vfield,fname,ii,999,0,2,nx,ierr)
         if (ierr[0]==1):
            exit(1)
# display pressure tensor in real space
      if (npro[0] > 2):
         ufield[:,:] = fms[4:10,:]
         fname = numpy.str.rstrip(sname[n-1])+numpy.str.rstrip(ename[2])
         graf1.dvector1(ufield,fname,ii,999,0,2,nx,ierr)
         if (ierr[0]==1):
            exit(1)
# display electron heat flux in real space
      if (npro[0]==4):
         vfield[:,:] = fms[11:14,:]
         fname = numpy.str.rstrip(sname[n-1])+numpy.str.rstrip(ename[4])
         graf1.dvector1(vfield,fname,ii,999,0,2,nx,ierr)
         if (ierr[0]==1):
            exit(1)

cmfield1.closeff(iudm)
cmfield1.closeff(iuv)
graf1.close_graphs()
